lib/Sema/CSRanking.cpp

```diff
@@ -1188,6 +1188,21 @@ SolutionCompareResult ConstraintSystem::compareSolutions(
       continue;
     }
 
+    // Prefer binding to an l-value over an r-value of the same type, since we
+    // can load from an l-value to produce an r-value during solution
+    // application.
+    if (type1->is<LValueType>() != type2->is<LValueType>()) {
```

Comment:
This as a general rule doesn't make much sense to me. What are the situations this makes any deference?

Reply:
@xedin In the example we were discussing above:
```
let _: Foo = .optFoo!
```
we end up with the constraints:
- `$T2 conv $T3`
- `$T3 conv Foo`

Where `$T2` is the type of the `ForceValueExpr` and `$T3` is the type of our phantom `UnresolvedMemberChainResultExpr`. Since `optFoo` yields an lvalue, we get two solutions:
1. `$T2 as @lvalue Foo, $T3 as Foo`
2. `$T2 as @lvalue Foo, $T3 as @lvalue Foo`

Basically, the lvalue conversion can happen at either “step”, from `$T2` to `$T3`, or `$T3` to the contextual type. Under existing rules, this results in an ambiguity since neither solution is preferred to the other.

Either solution, once applied, would result in the same types in the resulting AST, so we really just need to pick one arbitrarily. Is this logic erroneous in some cases where we’ve ended up with two solutions differing only in a binding based on lvalue-ness?

Reply:
Since `$T3` is a result of the chain it should always be r-value. Result of the chain should always be loaded into resulting type variable if the last member is l-value (type variable shouldn’t allow l-value as well).

Reply:
@xedin We need to allow for lvalues in at least some cases. When I initially worked on this implementation I tried disallowing lvalues at the tail, but this code type-checks today:

```
struct Foo {
  static var foo: Foo = Foo()
}

let foo2 = Foo();

.foo = foo2
```
I suppose we could in theory disallow lvalues for multi-member chains, but that seems inconsistent and would require us to special-case the single-member situation during constraint generation

Reply:
Hm, I think it this case solver should only infer one binding for the "$T3" in your example which always preserves l-valueness... Do you know how did we end up with r-value type as a binding for it?

Reply:
Yeah, this might be a bit tricky to do, effectively `$T3` has to be r-value in that example because of direct subtype binding but it could also be l-value because it's allowed to implicitly load when necessary... What I wonder is whether it would be possible to avoid introducing "in-between" type variable here and just return a type of the "tail" and use it as a result type of the chain (base is going to be directly convertible to that)?... That would solve this l-value ambiguity problem, but there might be situations where this type effectively acted as join between type of the member and contextual type... Otherwise we might need a special locator element which would get detected in `matchTypes` and would disallow implicit loading of the value to make sure that "result" is always r-value if context is r-value...

Reply:
@xedin We need the "in between" variable to deal with the fact that the type of the last member is permitted to be a subtype of the contextual type of the chain, as in the `let _: Base = .sub` up-thread, where we want the type checked expression to be `let _: Base = Base.sub` rather than `let _: Base = Sub.sub`.

This PR already introduces the `UnresolvedMemberChainResult` locator type, and the type variable created to represent the result of the chain is located using that path element, so it should be possible to add logic to `matchTypes` where, if we're dealing with a type variable representing a member chain result and a conversion to a contextual type which is not an lvalue, we disallow lvalue conversion. But that feels messier and more error-prone than resolving the ambiguity at the `compareSolutions` stage.

Reply:
Actually I guess it could be a bit simpler... the member -> result conversion uses the `UnresolvedMemberChainResult` locator type, so we could categorically refuse to perform lvalue conversion at the member -> result step. This would have the opposite effect of your suggestion, since the lvalue conversion would always happen at result -> context rather than member -> result, but the outcome is the same—we only ever perform lvalue conversion at one position, so the ambiguity gets resolved.

Reply:
I'd like to avoid adding any new general ranking rules because it usually causes impact in other places (we could localize that to only "result" of implicit chains but that still just seems like a hack) especially in artificial situations like this one. We should try to resolve this in the solver, it would be great if we could just remove that "result" variable but if we can't - either matching or binding inference (preferable) should be adjusted to resolve ambiguity.

Reply:
That makes total sense, thanks for explaining. I'm not seeing a good way to remove the "result" variable and still be able to type check examples like `let _: Base = .sub` or `let _: Base = .Sub()` properly (unless you have suggestions), so I think we'll have to work with the matching/binding logic. Will ping you when I have changes ready!

Reply:
Yeah, cases like that is what I called "join" between tail and context, I don't have any ideas at the moment, need to experiment with that a bit...

Reply:
@xedin I dug in deeper to track down the source of the ambiguous bindings, and realized that the "result" type var was actually inheriting bindings from logic that was meant to capture only `ForceValueExpr`. Since I was anchoring the result type var at the position of the last member, we applied the same binding logic which resulted in more bindings than we actually need to solve the system.

I just pushed a change which anchors the result type var at the `UnresolvedMemberChainResultExpr` instead (which seems more correct anyway) and added some additional tests for the lvalue behavior. Didn't need to touch the binding/matching logic at all!

Reply:
Ah great! I expected something like this...